cl/redisapp/state.go

```diff
@@ -0,0 +1,281 @@
+	}
+
+	s.logger.Info("Loaded leader block build state", "CurrentStep", state.CurrentStep.String())
+	s.blockBuildState = &state
+	return nil
+}
+
+func (s *RedisStateManager) SaveBlockState(ctx context.Context) error {
+	s.blockStateMutex.Lock()
+	defer s.blockStateMutex.Unlock()
+
+	data, err := json.Marshal(s.blockBuildState)
+	if err != nil {
+		return fmt.Errorf("failed to serialize leader block build state: %w", err)
+	}
+
+	if err := s.redisClient.Set(ctx, s.blockStateKey, data, 0).Err(); err != nil {
+		return fmt.Errorf("failed to save leader block build state to Redis: %w", err)
+	}
+
+	return nil
+}
+
+func (s *RedisStateManager) ResetBlockState(ctx context.Context) error {
+	s.blockStateMutex.Lock()
+	s.blockBuildState = &types.BlockBuildState{
+		CurrentStep: types.StepBuildBlock,
+	}
+	s.blockStateMutex.Unlock()
+
+	if err := s.SaveBlockState(ctx); err != nil {
+		return fmt.Errorf("failed to reset leader state: %w", err)
+	}
+
+	return nil
+}
+
+func (s *RedisStateManager) SaveExecutionHeadAndAck(ctx context.Context, head *types.ExecutionHead, messageID string) error {
+	data, err := json.Marshal(head)
+	if err != nil {
+		return fmt.Errorf("failed to serialize execution head: %w", err)
+	}
+
+	key := fmt.Sprintf("executionHead:%s", s.InstanceID)
+	pipe := s.redisClient.TxPipeline()
+
+	pipe.Set(ctx, key, data, 0)
+	pipe.XAck(ctx, redisStreamName, s.groupName, messageID)
+
+	if _, err := pipe.Exec(ctx); err != nil {
+		return fmt.Errorf("transaction failed: %w", err)
+	}
+
+	s.logger.Info("Follower: Execution head saved and message acknowledged", "MessageID", messageID)
+	return nil
+}
+
+func (s *RedisStateManager) SaveBlockStateAndPublishToStream(ctx context.Context, bsState *types.BlockBuildState) error {
+	s.blockStateMutex.Lock()
+	defer s.blockStateMutex.Unlock()
+
+	s.blockBuildState = bsState
+	data, err := json.Marshal(bsState)
+	if err != nil {
+		return fmt.Errorf("failed to serialize leader block build state: %w", err)
+	}
+
+	pipe := s.redisClient.Pipeline()
+	pipe.Set(ctx, s.blockStateKey, data, 0)
+
+	message := map[string]interface{}{
+		"payload_id":         bsState.PayloadID,
+		"execution_payload":  bsState.ExecutionPayload,
+		"timestamp":          time.Now().UnixNano(),
+		"sender_instance_id": s.InstanceID,
+	}
+
+	pipe.XAdd(ctx, &redis.XAddArgs{
+		Stream: redisStreamName,
+		Values: message,
+	})
+
+	if _, err := pipe.Exec(ctx); err != nil {
+		return fmt.Errorf("pipeline failed: %w", err)
+	}
+
+	return nil
+}
+
+func (s *RedisStateManager) GetBlockBuildState(ctx context.Context) types.BlockBuildState {
+	s.blockStateMutex.Lock()
+	defer s.blockStateMutex.Unlock()
+
+	// Return a copy of the state to prevent external modification
+	return *s.blockBuildState
+}
+
+func (s *RedisStateManager) CreateConsumerGroup(ctx context.Context) error {
+	if err := s.redisClient.XGroupCreateMkStream(ctx, redisStreamName, s.groupName, "0").Err(); err != nil {
+		if !strings.Contains(err.Error(), "BUSYGROUP") {
```

Comment:
Doesn't the redis client export the errors that can be checked, instead of checking part of the error string?

Reply:
For this case no:

https://github.com/redis/go-redis/blob/1ed936eb097d10f3aaf40a44d282778f879eac32/commands_test.go#L6032

Reply:
https://github.com/redis/go-redis/issues/2265